scripts/build/gen_relocate_app.py

```diff
@@ -45,7 +45,7 @@
 
 
 PRINT_TEMPLATE = """
-                KEEP(*({0}))
+                *({0})
```

Comment:
although it does appear logical that we do not enforce `KEEP` on this, then i'm concerned for users using this functionality.

How do such users become aware that `KEEP` has now been removed, and how do they tell the build system they want to have `KEEP` added to those input sections ?

The script, including `KEEP` was originally introduced here: https://github.com/zephyrproject-rtos/zephyr/pull/10892
by @AdithyaBaglody 
@AdithyaBaglody would be nice if you can provide some background information for `KEEP`.

Or maybe @nashif knows something.

Temporary blocking to get some more background knowledge on this particular feature.

Reply:
Agreed that this PR should wait until the change gets some more feedback. Although I have not seen any adverse affects, there may be a reason that `KEEP` is required here.

Reply:
Its been a long time, i dont fully remember the issue i was facing. But what i can recall is that unless i added the KEEP the linker was removing some of the symbols because of optimization. The problem was that linker doesnt know that these symbols will be used elsewhere and it just removed  that symbol because it didn't find any reference for it in any compiled C file.

Reply:
I could see this being an issue if the relocated code was something relevant to the hardware (such as a vector table or bootloader block), but in that case wouldn't the need for `KEEP` be something external to the code relocation itself? Maybe the best way to handle this is adding a `KEEP` argument to the `zephyr_code_relocate` macro to allow the user to choose the behavior (since they can't otherwise control the presence of the `KEEP` flag when using `zephyr_code_relocate`) @tejlmand, what are your thoughts on this? The feature could be integrated with #50791, since that PR already proposes changing the API for `zephyr_code_relocate`

Reply:
>  @tejlmand, what are your thoughts on this? 

I believe the possibility to request the `zephyr_code_relocate()` to apply `KEEP`.
My biggest concern in this regard is that is that users who relies on `KEEP` being added doesn't know about this new argument.
And the missing `KEEP` will be one of those hard to debug issues.

What we can do is to introduce two flags `KEEP` and `NO_KEEP`, `zephyr_code_relocate(... [KEEP|NO_KEEP])` and if no flag is given then print a CMake warning that one of those flags should be used but still default to `KEEP`.
Then after two releases fail if no flag is provided.

Finally, we can then make `NO_KEEP` the default if no flag is given.

Sure will take some releases, but I think changing current behavior can end in users that spend quite some time to debug why things no longer works.

Open to other proposals.

Reply:
> Finally, we can then make `NO_KEEP` the default if no flag is given.
> 
> Sure will take some releases

What about having `KEEP` as default to not break anything?

Reply:
> What about having KEEP as default to not break anything?

`KEEP` would still be the default for now (and minimum two releases in future).
This was a comment regarding the wish to not have `KEEP` the default.

To generally ensure smallest rom usage as well as the assumption that most users might want to not use `KEEP`, then this is how a slowly transition can be made.

I assume that only special sections will need the `KEEP`.